### PR TITLE
Created server with endpoints and json file // Monday 10/31 homework

### PR DESCRIPTION
**Pull request by gilmored**

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	for _, item := range products {
+		//if string(item.ID) == params["id"] {
+		if strconv.Itoa(item.ID) == params["id"] {
+			json.NewEncoder(w).Encode(item)
+			return
+		}
+	}
+	json.NewEncoder(w).Encode(&product{})
```
Comment (jboursiquot) [inline cmd/server/main.go]: Why encode an empty product here?

File: cmd/server/main.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+func createProduct(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	var prod product
+	_ = json.NewDecoder(r.Body).Decode(&prod)
+	prod.ID = rand.Intn(10000000)
```
Comment (jboursiquot) [inline cmd/server/main.go]: Without a proper seed, `rand.Intn` will yield deterministic values. See https://gobyexample.com/random-numbers

File: cmd/server/main.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	json.NewEncoder(w).Encode(prod)
+
+}
+
+func updateProduct(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	for index, item := range products {
+		if strconv.Itoa(item.ID) == params["id"] {
+			products = append(products[:index], products[index+1:]...)
```
Comment (jboursiquot) [inline cmd/server/main.go:77]: What are you looking to accomplish with this operation? You're appending to `products` on line 92 anyways.

File: cmd/server/main.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+			var prod product
+			_ = json.NewDecoder(r.Body).Decode(&prod)
+			prod.ID, _ = strconv.Atoi(params["id"])
```
Comment (jboursiquot) [inline cmd/server/main.go]: Do not discard the error value. Handle it instead.

**Review by jboursiquot — CHANGES_REQUESTED** [review]
Good first pass. Address the suggestions and you'll be good to go.

File: cmd/server/main.go [hunk 5]
```diff
@@ -83,12 +83,11 @@ func createProduct(w http.ResponseWriter, r *http.Request) {
 func updateProduct(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	params := mux.Vars(r)
+	var prod product
 	for index, item := range products {
 		if strconv.Itoa(item.ID) == params["id"] {
 			products = append(products[:index], products[index+1:]...)
-			var prod product
 			_ = json.NewDecoder(r.Body).Decode(&prod)
```
Comment (jboursiquot) [inline cmd/server/main.go]: Handle your errors

File: cmd/server/main.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		log.Fatal(err)
+	}
+}
+
+func main() {
+	initProducts()
+	r := mux.NewRouter()
+	r.HandleFunc("/products", getProductsHandler)
+
+	r.HandleFunc("/api/products", getProducts).Methods("GET")
```
Comment (jboursiquot) [inline cmd/server/main.go]: Test your server and ensure you can reach all the endpoints from a client. Order of routes matter.

**Review by jboursiquot — CHANGES_REQUESTED** [review]
Some changes needed.

File: cmd/server/main.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	if err := json.Unmarshal(bs, &products); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func main() {
+	initProducts()
+	r := mux.NewRouter()
+	r.HandleFunc("/products", getProductsHandler)
```
Comment (jboursiquot) [inline cmd/server/main.go]: Why two products handlers that do the same thing?

File: cmd/server/main.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+				return
+			}
+		}
+	}
+}
+
+func createProduct(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	var prod product
+	//_ = json.NewDecoder(r.Body).Decode(&prod)
```
Comment (jboursiquot) [inline cmd/server/main.go]: Avoid leaving commented out code behind.

File: cmd/server/main.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+	if err := json.NewDecoder(r.Body).Decode(&prod); err != nil {
+		log.Println("error", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	prod.ID = rand.Intn(int(time.Now().UnixNano()))
```
Comment (jboursiquot) [inline cmd/server/main.go]: Without proper seeding this will not be random. See https://gobyexample.com/random-numbers

File: cmd/server/main.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+}
+
+func updateProduct(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	var prod product
+	id, err := strconv.Atoi(params["id"])
+	for index, item := range products {
+		if err != nil {
```
Comment (jboursiquot) [inline cmd/server/main.go]: You should do this check before starting the loop.

File: cmd/server/main.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	json.NewEncoder(w).Encode(products)
+}
+
+func deleteProduct(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	id, err := strconv.Atoi(params["id"])
+	for index, item := range products {
+		if err != nil {
```
Comment (jboursiquot) [inline cmd/server/main.go]: Check the err before the loop.

**Review by jboursiquot — CHANGES_REQUESTED** [review]
Some changes needed. Almost there. Don't forget to test the order of your routes to ensure you can reach every endpoint while the server's running.

File: cmd/server/main.go (hunk 7 above)
Comment (gilmored) [reply cmd/server/main.go]: Working through these now. Adding some comments to help me keep up with where I am. 

Oversight, removed.

File: cmd/server/main.go (hunk 8 above)
Comment (gilmored) [reply cmd/server/main.go]: Removed

File: cmd/server/main.go (hunk 10 above)
Comment (gilmored) [reply cmd/server/main.go]: Moved error check

File: cmd/server/main.go (hunk 11 above)
Comment (gilmored) [reply cmd/server/main.go]: Moved error check

File: cmd/server/main.go (hunk 9 above)
Comment (gilmored) [reply cmd/server/main.go]: Added seeding and checks to ensure there isn't duplication of IDs

File: cmd/server/main.go [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+				return
+			}
+		}
+	}
+}
+
+func createProduct(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	var prod product
+	isID := false
```
Comment (jboursiquot) [inline cmd/server/main.go]: You might be overthinking obtaining a new random ID. 
All you need is something like this:

```go
s1 := rand.NewSource(time.Now().UnixNano())
r1 := rand.New(s1)
r1.Intn(100)
```

**Review by jboursiquot — CHANGES_REQUESTED** [review]
Small change needed and you're there.

**Pull request merged**